### DISCUSSION TITLE
Propagating a tensor "v0" from First Layer to Subsequent Layers in Megatron with PP and TP

### DISCUSSION THREAD

**Discussion by shiroko98** [NVIDIA/Megatron-LM#1522] [Q&A]
Hi all,

I’m working on a custom model in the Megatron framework and need help with propagating a tensor, v0, computed in the first layer to all subsequent layers. The model runs with both pipeline parallelism (PP) and tensor parallelism (TP), and I’m unsure how to handle this in a distributed setting.

Goal:
Compute v0 in the first layer (e.g., when layer_number == 1).
Make v0 available to all later layers for use in their computations, while preserving gradients for backpropagation.

Questions:
What’s the recommended approach to pass v0 across all later layers in Megatron’s pipeline parallelism?
How can I efficiently share v0 across PP and TP ranks?
Any tips for ensuring gradients are correctly handled in this setup?

The followings are the corresponding formulas:
![1](https://github.com/user-attachments/assets/d5977af7-83c0-44e1-9869-6f3974484e9f)
![2](https://github.com/user-attachments/assets/be6a7344-2243-4440-9207-83ba8b45b1e7)

I’d appreciate any high-level guidance, examples, or references to relevant parts of the Megatron codebase. Thanks!

**Comment by Advait251206** [comment]
In Megatron-LM with **Pipeline Parallelism (PP)** and **Tensor Parallelism (TP)**, propagating a tensor like `v0` (computed in the first layer) to all subsequent layers is possible, but you must align with how Megatron already propagates **activations across pipeline stages** and **sharded tensors across TP ranks**.

The key challenge is:

> PP splits execution across stages (layers live on different GPUs), while TP shards tensors inside each layer.  
So `v0` must be both **pipeline-safe** and **tensor-parallel-safe**, while still remaining in the autograd graph.

---

# 1. Correct mental model

You want:

- Compute `v0` once (layer 1)
- Reuse it in all later layers
- Preserve gradients
- Work under PP + TP

So effectively:

```
Layer 1:
    compute v0
    return (hidden_state, v0)

Pipeline:
    stage0 → stage1 → stage2 → ...

Each stage:
    receives same v0
    uses v0 in computation
```

---

# 2. Pipeline Parallelism solution (Megatron-native way)

Megatron already supports passing extra tensors through pipeline stages.

### Step 1: Return `v0` from the first layer

```python
def forward(self, x, v0=None):

    if self.layer_number == 1:
        v0 = self.compute_v0(x)

    return self.block(x, v0=v0), v0
```

---

### Step 2: Thread `v0` through all layers

Every layer must accept and return it:

```python
def forward(self, x, v0=None):
    x = self.block(x, v0=v0)
    return x, v0
```

---

### Step 3: Ensure PP passes it across stages

Megatron pipeline execution already does:

```
output_stage_i → input_stage_{i+1}
```

So as long as `v0` is part of the forward output tuple, it will be automatically transferred across pipeline boundaries.

---

# 3. Tensor Parallelism handling (critical part)

Inside each layer, TP shards activations and weights.

So `v0` must be made TP-consistent.

---

## Option A (recommended): replicate `v0` across TP ranks

```python
from megatron.core.tensor_parallel.mappings import (
    copy_to_tensor_model_parallel_region
)

v0_tp = copy_to_tensor_model_parallel_region(v0)
```

### Why this is preferred:
- All TP ranks see identical `v0`
- No need to modify downstream math
- Simplifies debugging

---

## Option B: shard `v0` across TP ranks (advanced)

```python
from megatron.core.tensor_parallel.mappings import (
    scatter_to_tensor_model_parallel_region
)

v0_tp = scatter_to_tensor_model_parallel_region(v0)
```

Use only if:
- `v0` is very large
- downstream ops are TP-aware

---

# 4. Gradient preservation rules

To ensure correct backpropagation:

### MUST:
- Keep `v0` as a torch tensor
- Keep it inside forward graph
- Avoid in-place modifications

### MUST NOT:
- `.detach()`
- `.item()`
- storing in global variables
- converting to numpy

---

# 5. Clean Megatron-style implementation pattern

This is the most robust pattern:

```python
def forward(self, x, v0=None):

    # compute once
    if v0 is None:
        v0 = self.compute_v0(x)

    # ensure TP consistency
    v0 = copy_to_tensor_model_parallel_region(v0)

    # use in block
    x = self.block(x, v0=v0)

    return x, v0
```

Pipeline automatically carries `(x, v0)` forward.

---

# 6. Where people usually go wrong

### ❌ Mistake 1: storing v0 outside forward
Breaks autograd + PP tracking

---

### ❌ Mistake 2: computing v0 per layer
Kills correctness (you want shared signal)

---

### ❌ Mistake 3: ignoring TP
Leads to shape mismatch or silent divergence

---

### ❌ Mistake 4: assuming PP shares tensors automatically
Only works if explicitly returned

---

# 7. Efficiency considerations

If `v0` is reused in every layer:

### Good:
- compute once
- broadcast once
- reuse everywhere

### Better optimization:
If `v0` is constant per microbatch:
- compute in first pipeline stage only
- reuse cached value across forward passes

---

# 8. Summary

To propagate `v0` in Megatron with PP + TP:

### Pipeline Parallelism:
- Return `v0` from first layer
- Thread it through all layer forward calls
- Megatron automatically passes it between stages

### Tensor Parallelism:
- Use `copy_to_tensor_model_parallel_region(v0)` for replication
- Or scatter if memory optimization is needed

### Autograd:
- Keep everything inside torch graph
- Do not detach or convert

---

# 9. One-line answer

> You propagate `v0` in Megatron by **returning it through the pipeline forward pass and ensuring TP consistency using `copy_to_tensor_model_parallel_region`, so it flows across PP stages while remaining identical across TP ranks and fully differentiable.**

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
